File: frontend/src/api/glossary.ts
import api from "./client";

export interface GlossaryTerm {
  id: string;
  name: string;
  definition: string;
  tags?: string[];
  status: "draft" | "approved";
  owner_id?: string;
  owner_name?: string;
  created_by?: string;
  creator_name?: string;
  created_at: string;
  updated_at: string;
}

export interface TermLink {
  id: string;
  term_id: string;
  entity_type: string;
  entity_id: string;
  created_by?: string;
  created_at: string;
}

export interface PaginatedGlossary {
  total: number;
  page: number;
  size: number;
  items: GlossaryTerm[];
}

export const getGlossaryTerms = (page = 1, size = 20, q?: string) =>
  api.get<PaginatedGlossary>("/api/v1/glossary", { params: { page, size, ...(q ? { q } : {}) } }).then((r) => r.data);

export const getGlossaryTerm = (id: string) =>
  api.get<GlossaryTerm>(`/api/v1/glossary/${id}`).then((r) => r.data);

export const createGlossaryTerm = (data: { name: string; definition: string; tags?: string[]; status?: string }) =>
  api.post<GlossaryTerm>("/api/v1/glossary", data).then((r) => r.data);

export const patchGlossaryTerm = (id: string, data: Partial<{ name: string; definition: string; tags: string[]; status: string }>) =>
  api.patch<GlossaryTerm>(`/api/v1/glossary/${id}`, data).then((r) => r.data);

export const deleteGlossaryTerm = (id: string) =>
  api.delete(`/api/v1/glossary/${id}`);

export const getTermLinks = (termId: string) =>
  api.get<TermLink[]>(`/api/v1/glossary/${termId}/links`).then((r) => r.data);

export const linkTerm = (termId: string, entity_type: string, entity_id: string) =>
  api.post<TermLink>(`/api/v1/glossary/${termId}/links`, { entity_type, entity_id }).then((r) => r.data);

export const unlinkTerm = (termId: string, linkId: string) =>
  api.delete(`/api/v1/glossary/${termId}/links/${linkId}`);

File: backend/app/schemas/glossary.py
import uuid
from datetime import datetime

from pydantic import BaseModel


class GlossaryTermCreate(BaseModel):
    name: str
    definition: str
    tags: list[str] | None = None
    status: str = "draft"


class GlossaryTermPatch(BaseModel):
    name: str | None = None
    definition: str | None = None
    tags: list[str] | None = None
    status: str | None = None


class TermLinkOut(BaseModel):
    id: uuid.UUID
    entity_type: str
    entity_id: str
    created_at: datetime

    model_config = {"from_attributes": True}


class GlossaryTermOut(BaseModel):
    id: uuid.UUID
    name: str
    definition: str
    owner_id: uuid.UUID | None
    owner_name: str | None = None
    tags: list[str] | None
    status: str
    created_by: uuid.UUID | None
    creator_name: str | None = None
    created_at: datetime
    updated_at: datetime
    links: list[TermLinkOut] = []

    model_config = {"from_attributes": True}


class PaginatedGlossaryTerms(BaseModel):
    total: int
    page: int
    size: int
    items: list[GlossaryTermOut]


class TermLinkCreate(BaseModel):
    entity_type: str
    entity_id: str
